Add controller id range to ceph-nvmeof.conf and pass to gateway

gbregman opened the pull request:
Add controller id range to ceph-nvmeof.conf and pass to gateway. This allows using the partioning of controller id range between gateways, to be done later by Ceph-adm.

Related to issue #121

gbregman committed:
Add min and max controller id to ceph-nvmeof.conf and pass them to gateway. This allows using the partioning of controller
id range between gateways.

Signed-off-by: Gil Bregman <gbregman@il.ibm.com>

baum commented on ceph-nvmeof.conf:
```diff
@@ -15,6 +15,8 @@ port = 5500
 state_update_notify = True
 state_update_interval_sec = 5
+min_cntlid = 1
+max_cntlid = 65519
```
Hello @gbregman 🖖. Why 65519? Would expect 65535 or 0xffff

gbregman replied:
Alex, the current value in SPDK is 0xFFEF, which is 65519. I wanted to keep the current behavior until we set real values in the future using ceph-adm.

trociny replied:
May be use some more human readable names, like `max/min_controller_id`?
And we seemed to have an agreement to have config params that have default values to be commented out in the sample config.

gbregman replied:
Will rename the fields as suggested.

baum reviewed (approved):
lgtm 🖖

idryomov commented on control/grpc.py:
```diff
@@ -118,12 +118,16 @@ def create_subsystem(self, request, context=None):
         self.logger.info(
             f"Received request to create subsystem {request.subsystem_nqn}")
+        min_cntlid = self.config.getint_with_default("gateway", "min_cntlid", 1)
+        max_cntlid = self.config.getint_with_default("gateway", "max_cntlid", 65519)
```
Do we want to duplicate SPDK defaults here?  Perhaps a better approach would be to keep the current behavior of not passing anything in the RPC if the respective option is not provided?  @PepperJo Any preferences?

PepperJo replied:
No preference. Maybe we should print a warning that if not specified in a multi GW group multipathing will not work correctly. Alternatively we want to be more verbose we can also add the option to the omap and let the GW check that it doesn't overlap with any others and fail if it does.

gbregman replied:
@PepperJo I was told that the Ceph-adm will take care of managing the ranges between the different gateways so we just need a way for it to pass the values.

PepperJo replied:
Ok that makes sense. No need to get complicated then.

gbregman replied:
@PepperJo failing the gateway seems like an overkill. Especially when there is an easy workaround we used here to test two gateways. We just created a connection and the deleted it on one gateway. The result was that now one gateway advanced the controller id counter so now we could create a connection on both gateways and each one got a different controller id.

PepperJo replied:
I mean it's certainly not something a user should experience. But if we transfer this responsibility to cephadm I don't have any objections to not checking.

PepperJo reviewed (approved):
LGTM

idryomov commented:
@gbregman The commit message needs to be amended according to https://github.com/ceph/ceph-nvmeof/blob/devel/CONTRIBUTING.md#commit-format.  In particular, the title should be short and separated from the body.

idryomov commented:
If you see GitHub truncating the commit title and showing `...` there (in this case it also affects the PR title as there is just a single commit), it's good a sign that something is wrong ;)

gbregman committed:
Fixes #121: Add controller id range to ceph-nvmeof.conf and pass to gateway.
This allows using the partioning of controller id range between gateways,
to be done later by Ceph-adm.

Signed-off-by: Gil Bregman <gbregman@il.ibm.com>

gbregman committed:
control: add min and max controller id options

Add min and max controller id options to ceph-nvmeof.conf and pass
them to SPDK. This allows to partion the controller id range between
several gateways in a gateway group.

Fixes: #121

Signed-off-by: Gil Bregman <gbregman@il.ibm.com>

idryomov commented on ceph-nvmeof.conf:
```diff
@@ -15,8 +15,8 @@ port = 5500
-min_cntlid = 1
-max_cntlid = 65519
+#min_controller_id = 1
+#max_controller_id = 65519
```
No other option is commented out, let's get rid of `#` here.

gbregman replied:
But according to @trociny the standard is to comment out fields which are set to the default.

idryomov replied:
Ah, I forgot that we have made that change.  Apologies for the noise!

idryomov commented:
@gbregman Please squash the second commit so that there is only one.  The commit message still needs to be adjusted too.  I would suggest:
```
control: add min and max controller id options

Add min and max controller id options to ceph-nvmeof.conf and pass
them to SPDK. This allows to partion the controller id range between
several gateways in a gateway group.

Fixes: #121

Signed-off-by: Gil Bregman <gbregman@il.ibm.com>
```

gbregman committed:
Merge branch 'controller-range' of https://github.com/ceph/ceph-nvmeof into controller-range

idryomov commented:
> Please squash the second commit so that there is only one.

@gbregman This ended up with 4 commits, one of which is a bogus merge commit (while the "real" merge commit is missing `Reviewed-by` tags).  Please don't merge PRs with outstanding comments even if the comment is not on the code itself.
